### PR TITLE
Add Google Artifact Registry support to java_export

### LINKED ISSUE #774: java_export: support pushing to GCP Artifact Registry
It'd be great if `java_export` can push to `artifactregistry://`, in addition to S3 (#545) and GCS(#458)

https://github.com/GoogleCloudPlatform/artifact-registry-maven-tools

### PR DESCRIPTION
**Pull request by tsawada**
Fixes #774

This PR adds support to publishing to Google Artifact Registry `artifactregistry://` URLs.
https://cloud.google.com/artifact-registry

### CODE REVIEW

**Comment by tsawada** [comment]
As far as I understand, Artifact Registry is just a simple https server except that it requires OAuth2 tokens in `Authorization` header. I modified `httpUpload` function to accept `com.google.auth.Credentials` and converted existing User/Password credentials to also base on the `com.google.auth.Credentials` class.

I think GCS (and maybe even S3) works similarly and it is possible to remove those custom libraries with just a simple credentials handling + http PUT call. I don't see much benefit though.

**Comment by tsawada** [comment]
@shs96c can you take a look?

File: private/tools/java/rules/jvm/external/maven/MavenPublisher.java [hunk 1]
```diff
@@ -289,6 +303,14 @@ private static Callable<Void> s3upload(String targetUrl, Path toUpload) {
     };
   }
 
+  private static Callable<Void> arUpload(String targetUrl, Path toUpload) {
+    return () -> {
+      Credentials cred = GoogleCredentials.getApplicationDefault();
+      String url = "https://" + targetUrl.substring(19);
```
Comment (shs96c) [inline private/tools/java/rules/jvm/external/maven/MavenPublisher.java:309]: Presumably, `19` is the length of the scheme?

**Review by shs96c — APPROVED** [review]
Looks good to me. Thank you for putting this together!

File: private/tools/java/rules/jvm/external/maven/MavenPublisher.java (hunk 1 above)
Comment (tsawada) [reply private/tools/java/rules/jvm/external/maven/MavenPublisher.java:309]: yes, it's `len("artifactregistry://")`

**Comment by cognitiaclaeves** [comment]
Sorry to necro this PR. A company I'm working for is needing to use GAR in the opposite direction -- dependency imports. I've had difficulty finding any information out there on how to do use GAR with bazel. @tsawada, I'm wondering if you ever had a need to import what you exported, as you seemed to have enough of an interest to export to GAR that you contributed this code ...

**Pull request merged**